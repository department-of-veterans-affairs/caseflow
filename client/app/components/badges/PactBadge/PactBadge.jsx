import PropTypes from 'prop-types';
import * as React from 'react';

import Badge from '../Badge';
import { COLORS } from 'app/constants/AppConstants';

/**
 * Component to display if the appeal is a pact.
 */

const PactBadge = (props) => {
  const { appeal } = props;

  if (!appeal.pact) {
    return null;
  }

  const tooltipText = "Appeal has issue(s) related to Promise to Address Comprehensive Toxics (PACT) Act.";

  return <Badge
    name="pact"
    displayName="PACT"
    color={COLORS.ORANGE}
    tooltipText={tooltipText}
    id={`pact-${appeal.id}`}
    ariaLabel="pact"
  />;
};

PactBadge.propTypes = {
  appeal: PropTypes.object
};

export default PactBadge;
