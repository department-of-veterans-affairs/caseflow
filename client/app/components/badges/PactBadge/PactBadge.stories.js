import React from 'react';

import PactBadgeComponent from './PactBadge';

export default {
  title: 'Commons/Components/Badges/PACT Badge',
  component: PactBadgeComponent,
  parameters: {
    layout: 'centered',
  },
  args: {
    appeal: {
      pact: true,
    }
  }
};

const Template = (args) => <PactBadgeComponent {...args} />;

export const PactBadge = Template.bind({});
