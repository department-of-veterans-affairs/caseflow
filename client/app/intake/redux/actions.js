import { ACTIONS } from '../constants';
import ApiUtil from '../../util/ApiUtil';
import { formatDateStringForApi } from '../../util/DateUtil';

export const startNewIntake = () => ({
  type: ACTIONS.START_NEW_INTAKE
});

export const setFileNumberSearch = (fileNumber) => ({
  type: ACTIONS.SET_FILE_NUMBER_SEARCH,
  payload: {
    fileNumber
  }
});

export const doFileNumberSearch = (fileNumberSearch) => (dispatch) => {
  dispatch({
    type: ACTIONS.FILE_NUMBER_SEARCH_START
  });

  return ApiUtil.post('/intake', { data: { file_number: fileNumberSearch } }).
    then(
      (response) => {
        const responseObject = JSON.parse(response.text);

        dispatch({
          type: ACTIONS.FILE_NUMBER_SEARCH_SUCCEED,
          payload: {
            intakeId: responseObject.id,
            name: responseObject.veteran_name,
            formName: responseObject.veteran_form_name,
            fileNumber: responseObject.veteran_file_number
          }
        });
      },
      (error) => {
        const responseObject = JSON.parse(error.response.text);

        dispatch({
          type: ACTIONS.FILE_NUMBER_SEARCH_FAIL,
          payload: {
            errorCode: responseObject.error_code
          }
        });
      }
    );
};

export const setSelectedOption = (optionSelected) => ({
  type: ACTIONS.SET_SELECTED_OPTION,
  payload: {
    optionSelected
  }
});

export const setReceiptDate = (receiptDate) => ({
  type: ACTIONS.SET_RECEIPT_DATE,
  payload: {
    receiptDate
  }
});

export const submitReview = (rampElection) => (dispatch) => {
  dispatch({
    type: ACTIONS.SUBMIT_REVIEW_START
  });

  const data = {
    option_selected: rampElection.optionSelected,
    receipt_date: formatDateStringForApi(rampElection.receiptDate)
  };

  return ApiUtil.patch(`/intake/ramp/${rampElection.intakeId}`, { data }).
    then(
      () => dispatch({ type: ACTIONS.SUBMIT_REVIEW_SUCCEED }),
      () => dispatch({ type: ACTIONS.SUBMIT_REVIEW_FAIL })
    );
};
