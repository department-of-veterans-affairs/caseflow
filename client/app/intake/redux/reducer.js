import { ACTIONS, REQUEST_STATE } from '../constants';
import { update } from '../../util/ReducerUtil';

export const mapDataToInitialState = (data = {}) => ({
  veteran: {
    name: null,
    formName: null,
    fileNumber: null
  },
  inputs: {
    fileNumberSearch: '',
    receiptDateStr: '',
    veteranResponse: null
  },
  requestStatus: {
    fileNumberSearch: REQUEST_STATE.NOT_STARTED,
    submitReview: REQUEST_STATE.NOT_STARTED
  },
  rampElection: {
    intakeId: data.intakeId,
    optionSelected: null,
    receiptDate: null
  },
  searchError: null
});

// The keys in this object need to be snake_case
// because they're being matched to server response values.
const searchErrors = {
  invalid_file_number: {
    title: 'Veteran ID not found',
    body: 'Please enter a valid Veteran ID and try again.'
  },
  veteran_not_found: {
    title: 'Veteran ID not found',
    body: 'Please enter a valid Veteran ID and try again.'
  },
  veteran_not_accessible: {
    title: 'You don\'t have permission to view this veteran\'s informationâ€‹',
    body: 'Please enter a valid Veteran ID and try again.'
  },
  did_not_receive_ramp_election: {
    title: 'No opt-in letter was sent to this veteran',
    body: "An opt-in letter was not sent to this Veteran, so this form can't be processed" +
      'Please enter a valid Veteran ID below.'
  },
  default: {
    title: 'Something went wrong',
    body: 'Please try again. If the problem persists, please contact Caseflow support.'
  }
};


export const reducer = (state = mapDataToInitialState(), action) => {
  switch (action.type) {
  case ACTIONS.START_NEW_INTAKE:
    return mapDataToInitialState();
  case ACTIONS.SET_FILE_NUMBER_SEARCH:
    return update(state, {
      inputs: {
        fileNumberSearch: {
          $set: action.payload.fileNumber
        }
      }
    });
  case ACTIONS.SET_SELECTED_OPTION:
    return update(state, {
      rampElection: {
        optionSelected: {
          $set: action.payload.optionSelected
        }
      }
    });
  case ACTIONS.SET_RECEIPT_DATE:
    return update(state, {
      rampElection: {
        receiptDate: {
          $set: action.payload.receiptDate
        }
      }
    });
  case ACTIONS.FILE_NUMBER_SEARCH_START:
    return update(state, {
      requestStatus: {
        fileNumberSearch: {
          $set: REQUEST_STATE.IN_PROGRESS
        }
      }
    });
  case ACTIONS.FILE_NUMBER_SEARCH_SUCCEED:
    return update(state, {
      requestStatus: {
        fileNumberSearch: {
          $set: REQUEST_STATE.SUCCEEDED
        }
      },
      veteran: {
        name: {
          $set: action.payload.name
        },
        formName: {
          $set: action.payload.formName
        },
        fileNumber: {
          $set: action.payload.fileNumber
        }
      },
      rampElection: {
        intakeId: {
          $set: action.payload.intakeId
        }
      }
    });
  case ACTIONS.FILE_NUMBER_SEARCH_FAIL:
    return update(state, {
      searchError: {
        $set: searchErrors[action.payload.errorCode] || searchErrors.default
      },
      requestStatus: {
        fileNumberSearch: {
          $set: REQUEST_STATE.FAILED
        }
      }
    });
  case ACTIONS.SUBMIT_REVIEW_START:
    return update(state, {
      requestStatus: {
        submitReview: {
          $set: REQUEST_STATE.IN_PROGRESS
        }
      }
    });
  case ACTIONS.SUBMIT_REVIEW_SUCCEED:
    return update(state, {
      requestStatus: {
        submitReview: {
          $set: REQUEST_STATE.SUCCEEDED
        }
      }
    });
  case ACTIONS.SUBMIT_REVIEW_FAIL:
    return update(state, {
      requestStatus: {
        submitReview: {
          $set: REQUEST_STATE.FAILED
        }
      }
    });
  default:
    return state;
  }
};
