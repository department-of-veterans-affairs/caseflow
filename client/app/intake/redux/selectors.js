import { createSelector } from 'reselect';
import { RAMP_INTAKE_STATES } from '../constants';

const getRampElection = (state) => state.rampElection;

export const getRampElectionStatus = createSelector(
  [getRampElection],
  (rampElection) => {
    if (rampElection.isComplete) {
      return RAMP_INTAKE_STATES.COMPLETED;
    } else if (rampElection.isReviewed) {
      return RAMP_INTAKE_STATES.REVIEWED;
    } else if (rampElection.intakeId) {
      return RAMP_INTAKE_STATES.STARTED;
    }

    return RAMP_INTAKE_STATES.NONE;
  }
);
