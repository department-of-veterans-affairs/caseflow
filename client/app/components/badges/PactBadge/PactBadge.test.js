import React from 'react';
import { mount } from 'enzyme';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware } from 'redux';

import rootReducer from 'app/queue/reducers';
import PactBadge from './PactBadge';

describe('PactBadge', () => {
  const defaultAppeal = {
    pact: true,
  };

  const getStore = () => createStore(rootReducer, applyMiddleware(thunk));

  const setupPactBadge = (store) => {
    return mount(
      <Provider store={store}>
        <PactBadge
          appeal={defaultAppeal}
        />
      </Provider>
    );
  };

  it('renders correctly', () => {
    const store = getStore();
    const component = setupPactBadge(store);

    expect(component).toMatchSnapshot();
  });
});
