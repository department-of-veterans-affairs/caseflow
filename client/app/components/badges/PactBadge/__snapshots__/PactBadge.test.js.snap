// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PactBadge renders correctly 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <PactBadge
    appeal={
      Object {
        "pact": true,
      }
    }
  />
</Provider>
`;
